### PR TITLE
New smart contract to batch escrow transactions

### PR DESCRIPTION
**Pull request by avclarke**
Enables us to escrow batches of addresses and tokens at a single time to speed up distribution.

### CODE REVIEW

File: contracts/EscrowDistributor.sol
Comment (JChiaramonte7) [inline contracts/EscrowDistributor.sol:33]: Spelling

File: contracts/EscrowDistributor.sol
Comment (JaredBorders) [inline contracts/EscrowDistributor.sol:43]: default value is zero. does not need to be explicitly set to zero

File: contracts/EscrowDistributor.sol
Comment (JaredBorders) [inline contracts/EscrowDistributor.sol:46]: can be optimized. See an example here: https://github.com/Kwenta/cc-vesting/blob/1f420f3fe7f342149214a488278cfca775eccaef/src/VestingFactory.sol#L129

Does not have to be a do-while but decrementing is cheaper and can be unchecked

File: contracts/EscrowDistributor.sol
Comment (JaredBorders) [inline contracts/EscrowDistributor.sol:61]: event params are not correct. Recommend the Event({paramName: name, (...) }) format

File: contracts/EscrowDistributor.sol
Comment (JaredBorders) [inline contracts/EscrowDistributor.sol:53]: can optimize (see above)

File: test/local/contracts/EscrowDistributor.test.ts
Comment (JaredBorders) [inline test/local/contracts/EscrowDistributor.test.ts:111]: Would be good to add test which ensure event params are correct

File: test/local/contracts/EscrowDistributor.test.ts
Comment (JaredBorders) [inline test/local/contracts/EscrowDistributor.test.ts:123]: would add tests for edge cases (i.e. when account address is zero or value is zero what happens?)

**Review by JaredBorders — CHANGES_REQUESTED** [review]
Left several comments! lmk if you have any questions.

File: contracts/EscrowDistributor.sol
Comment (JaredBorders) [inline contracts/EscrowDistributor.sol:10]: if you save this as IERC20 vs address, would that save gas? Would be good to know. That way you don't have to do the `IERC20(kwenta).transferFrom()` and can just do it `kwenta.transferFrom()`.

May not make a difference

File: test/local/contracts/EscrowDistributor.test.ts
Comment (avclarke) [reply test/local/contracts/EscrowDistributor.test.ts:111]: is there a different function for this, it seems `emit` only takes contract and eventName?

Comment (JaredBorders) [reply test/local/contracts/EscrowDistributor.test.ts:111]: Yes there is. Here is an example I have used in the past but there is also an easier way that is a little less verbose

https://github.com/Kwenta/margin-manager/blob/906267e512f4242274235800f3e03370b04d0e7d/test/integration/margin.behavior.ts#L1420

File: contracts/EscrowDistributor.sol
Comment (avclarke) [reply contracts/EscrowDistributor.sol:10]: Yea I did have it like that and then thought it might be handy to store address if ever anyone wanted to read which token address it was, but probably not really needed. Is there a simple way you compare gas, can you get that gas value when calling in a test?

File: test/local/contracts/EscrowDistributor.test.ts
Comment (JChiaramonte7) [reply test/local/contracts/EscrowDistributor.test.ts:111]: You can chain `emit()` with `withArgs(arg1, arg2, arg3...)`

https://ethereum-waffle.readthedocs.io/en/latest/matchers.html#emitting-events

File: contracts/EscrowDistributor.sol
Comment (JChiaramonte7) [reply contracts/EscrowDistributor.sol:10]: I believe even if you do store an address as a contract type, it will still return the address offchain (ie. on etherscan)

Example of StakingRewards if you try to read rewardEscrow or supplySchedule: https://goerli-optimism.etherscan.io/address/0x1653a3a3c4ccee0538685f1600a30df5e3ee830a#readContract

Comment (avclarke) [reply contracts/EscrowDistributor.sol:10]: ok good to know thanks, now storing both rewardEscrow and kwenta as contract instances instead of addresses

File: test/local/contracts/EscrowDistributor.test.ts
Comment (avclarke) [reply test/local/contracts/EscrowDistributor.test.ts:111]: perfect, thanks

File: contracts/EscrowDistributor.sol
Comment (JaredBorders) [reply contracts/EscrowDistributor.sol:10]: So preference I suppose. And as far as testing gas costs, hardhat has a plugin. Foundry reports it by default which is what i was using to test out my theory haha

**Review by JaredBorders — APPROVED** [review]
Looks good to me. I'd wait for Jeremy to review too before merging.

Nice work!!

File: contracts/EscrowDistributor.sol
Comment (avclarke) [reply contracts/EscrowDistributor.sol]: Good to know

**Review by JChiaramonte7 — APPROVED** [review]

**Pull request merged**
